pkg/downwardmetrics/vhostmd/disk.go

```diff
@@ -0,0 +1,196 @@
+package vhostmd
+
+import (
+	"encoding/binary"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics/vhostmd/api"
+)
+
+const fileSize = 262144
+const maxBodyLength = fileSize - 24
+
+var signature = [4]byte{'m', 'v', 'b', 'd'}
+
+type vhostmd struct {
+	filePath string
+}
+
+type Header struct {
+	Signature [4]byte
+	Flag      int32
+	Checksum  int32
+	Length    int32
+}
+
+type Disk struct {
+	Header *Header
+	Raw    []byte
+}
+
+func (d *Disk) String() string {
+	return fmt.Sprintf("%v:%v:%v:%v", string(d.Header.Signature[:]), d.Header.Flag, d.Header.Checksum, d.Header.Length)
+}
+
+func (d *Disk) Verify() error {
+	var checksum int32
+	for _, b := range d.Raw {
+		checksum = checksum + int32(b)
+	}
+	if d.Header.Flag > 0 {
+		return fmt.Errorf("file is locked")
+	}
+	if checksum != d.Header.Checksum {
+		return fmt.Errorf("checksum is %v, but expected %v", checksum, d.Header.Checksum)
+	}
+	return nil
+}
+
+func (d *Disk) Metrics() (*api.Metrics, error) {
+	m := &api.Metrics{}
+	if err := xml.Unmarshal(d.Raw, m); err != nil {
+		return nil, err
+	}
+	m.Text = strings.TrimSpace(m.Text)
+	for i, metric := range m.Metrics {
+		m.Metrics[i].Name = strings.TrimSpace(metric.Name)
+		m.Metrics[i].Type = api.MetricType(strings.TrimSpace(string(metric.Type)))
+		m.Metrics[i].Context = api.MetricContext(strings.TrimSpace(string(metric.Context)))
+		m.Metrics[i].Value = strings.TrimSpace(metric.Value)
+		m.Metrics[i].Text = strings.TrimSpace(metric.Text)
+	}
+	return m, nil
+}
+
+func (v *vhostmd) Create() error {
+	return createDisk(v.filePath)
+}
+
+func (v *vhostmd) Read() (*api.Metrics, error) {
+	disk, err := readDisk(v.filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load vhostmd file: %v", err)
+	}
+	if err := disk.Verify(); err != nil {
+		return nil, fmt.Errorf("failed to verify vhostmd file: %v", err)
+	}
+	return disk.Metrics()
+}
+
+func (v *vhostmd) Write(metrics *api.Metrics) error {
+	f, err := os.OpenFile(v.filePath, os.O_RDWR, 0)
+	if err != nil {
+		return fmt.Errorf("failed to open vhostmd disk: %v", err)
+	}
+	defer f.Close()
```

Comment:
Do we want to consider this here? https://www.joeshaw.org/dont-defer-close-on-writable-files/

Reply:
I can let it log an error. There is not action necessary regarding to recovering.

Reply:
Fixed.